spinalcordtoolbox/centerline/core.py

```diff
@@ -238,9 +239,21 @@ def get_centerline(im_seg, param=ParamCenterline(), verbose=1, remove_temp_files
                      else "centerline.nii.gz")
         im_centerline.save(os.path.join(tmp_folder, fname_ctr), mutable=True)
 
+    # If specified, adjust centerline coordinates (`Centerline.points`) and derivatives (`Centerline.derivatives`) to physical ("phys") space and native (`im_seg`) orientation
+    if space != 'pix':
+        # Transform centerline to physical coordinate system
+        arr_ctl = im_seg.transfo_pix2phys([[x_centerline_fit[i], y_centerline_fit[i], z_ref[i]]
+                                           for i in range(len(x_centerline_fit))]).T
+        # Adjust derivatives with pixel size
+        nx, ny, nz, nt, px, py, pz, pt = im_seg.change_orientation(native_orientation).dim
+        arr_ctl_der = np.array([x_centerline_deriv * px, y_centerline_deriv * py, np.ones_like(z_ref) * pz])
+    else:
+        arr_ctl = np.array([x_centerline_fit, y_centerline_fit, z_ref])
+        arr_ctl_der = np.array([x_centerline_deriv, y_centerline_deriv, np.ones_like(z_ref)])
+
     return (im_centerline,
-            np.array([x_centerline_fit, y_centerline_fit, z_ref]),
-            np.array([x_centerline_deriv, y_centerline_deriv, np.ones_like(z_ref)]),
+            arr_ctl,
+            arr_ctl_der,
```

Comment:
So, for example, in a call like this: 

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7cb30ddbaa88ded9c3ae6250c72832f5c40f4d64/spinalcordtoolbox/scripts/sct_analyze_lesion.py#L429-L430

We would replace `arr_ctl` and `arr_ctl_der` with a single `centerline` object that has an `arr_ctl_der` property:

```python
 _, centerline, _ = get_centerline(im_seg, param=ParamCenterline(), verbose=1) 
 x_centerline_deriv, y_centerline_deriv, z_centerline_deriv = centerline.arr_ctl_der 
```

And for a call like this:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7cb30ddbaa88ded9c3ae6250c72832f5c40f4d64/spinalcordtoolbox/straightening.py#L152

We would replace it with:

```python
 _, centerline, _ = get_centerline(img_ctl, self.param_centerline, verbose, remove_temp_files) 
```

Thus serving both purposes with a single `get_centerline` function.

Reply:
Yeah that would make a lot of sense. I am guessing that there is an automatic way to find every instance of
```
 _, arr_ctl, arr_ctl_der, _ = get_centerline(im_seg, param=ParamCenterline(), verbose=1)
```
and replace each one with
```
 _, centerline,_ = get_centerline(im_seg, param=ParamCenterline(), verbose=1)
arr_ctl = centerline.arr_ctl
arr_ctl_der = centerline.arr_ctl_der
```
Is this an error-prone process? The `centerline.core.get_centerline` function is quite ubiquitous (unlike `straightening._get_centerline()`, which I have only personally only encountered in the 2 places described in the issue [4006](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4006)).

Reply:
> I am guessing that there is an automatic way to find every instance of [old cpde] and replace each one with [new code]

My IDE of choice is [PyCharm](https://www.jetbrains.com/pycharm/), and technically they have a way to do Find+Replace over multiple lines using Regex patterns.

But, as you say, doing this automatically would likely be error-prone? (Especially since there are some cases where only 0/1 of `{arr_ctl,arr_ctl_der}` are used, rather than both.)

So my first instinct would be to do this replacement work manually by searching for the more simple `"= get_centerline("` term. This returns 19 instances for me:

<details>

![image](https://user-images.githubusercontent.com/16181459/217009047-099881bf-63bf-414f-9af9-2bd2cdb99b88.png)

</details>

Reply:
(But of course, we would first need to update the `Centerline` class to store the arrays, as well as update the `get_centerline` function to produce a `Centerline` object.)

Reply:
Hey! Sorry for the delay on this. Doesn't the Centerline class already store the arrays? (`centerline.points` and `centerline.derivatives`)? So I made a mistake above, and after updating `get_centerline` to produce a `Centerline` object, I will replace this:

` _, arr_ctl, arr_ctl_der, _ = get_centerline(im_seg, param=ParamCenterline(), verbose=1)`

with this:

```
 _, centerline,_ = get_centerline(im_seg, param=ParamCenterline(), verbose=1)
arr_ctl = centerline.points
arr_ctl_der = centerline.derivatives
```

If this seems correct to you, I will move forward with this (and double check the shapes of everything along the way to make sure not to miss any transposes)

Reply:
I have tried to implement the above plan in the commits below. :smile:

Reply:
You're absolutely right that `points` and `derivatives` are basically the same as `arr_ctl` and `arr_ctl_der`. Except, they're actually transposes of one another? 

So, in [1da5a3c](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4010/commits/1da5a3c25718e427b7e6f181e49fa52db393c049), I added properties that essentially just call `.T` (transpose) on the `points` and `derivatives` arrays to produce `arr_ctl` and `arr_ctl_der`.